src/tangent.jl

```diff
@@ -78,33 +78,104 @@ function Base.getindex(a::AbstractTangentBundle, b::TaylorTangentIndex)
     error("$(typeof(a)) is not taylor-like. Taylor indexing is ambiguous")
 end
 
+abstract type AbstractTangentSpace; end
+
+"""
+    struct ExplicitTangent{P}
+
+A fully explicit coordinate representation of the tangent space,
+represented by a vector of `2^(N-1)` partials.
+"""
+struct ExplicitTangent{P <: Tuple} <: AbstractTangentSpace
+    partials::P
+end
+
+"""
+    struct TaylorTangent{C}
+
+The taylor bundle construction mods out the full N-th order tangent bundle
+by the equivalence relation that coefficients of like-order basis elements be
+equal, i.e. rather than a generic element
+
+    a + b ∂₁ + c ∂₂ + d ∂₃ + e ∂₂ ∂₁ + f ∂₃ ∂₁ + g ∂₃ ∂₂ + h ∂₃ ∂₂ ∂₁
+
+we have a tuple (c₀, c₁, c₂, c₃) corresponding to the full element
+
+    c₀ + c₁ ∂₁ + c₁ ∂₂ + c₁ ∂₃ + c₂ ∂₂ ∂₁ + c₂ ∂₃ ∂₁ + c₂ ∂₃ ∂₂ + c₃ ∂₃ ∂₂ ∂₁
+
+i.e.
+
+    c₀ + c₁ (∂₁ + ∂₂ + ∂₃) + c₂ (∂₂ ∂₁ + ∂₃ ∂₁ + ∂₃ ∂₂) + c₃ ∂₃ ∂₂ ∂₁
+
+
+This restriction forms a submanifold of the original manifold. The naming is
+by analogy with the (truncated) Taylor series
```

Comment:
I must admit, I am still a little confused about the name `TaylorTangent` and why it's useful. That's mostly unrelated to this PR though.

Reply:
It implements Taylor mode AD like https://github.com/JuliaDiff/TaylorSeries.jl.

Reply:
I thought it still had separate coefficients for each variable, or did I misunderstand what the `∂ₙ` are? https://github.com/JuliaDiff/TaylorSeries.jl#examples

Reply:
This docstring shows the analogy between the `∂ₙ` and the coefficients of a taylor series. In particular, taylor mode takes advantage of symmetry of the derivative operators (i.e. the order of the operators is not distinguishable). For example, you can't use taylor mode to evaluate the second derivative in `∂₁ sin(∂₂ sin(x))`, but you can in `∂^2 sin(sin(x)) =  ∂₁ ∂₂ sin(sin(x)) = ∂₂ ∂₁`. Because of this symmetry under exchange of the operators, you also get symmetry in the tangent space that you can then represent explicitly.